services/twilioService.js

```diff
@@ -0,0 +1,225 @@
+      if (channel.createdBy === this.id) {
+        this.tryInviteContact(contact);
+      }
+
+      return true;
+    }
+    return false;
+  }
+
+  async getChannelForChat(contact) {
+    const channelName = this.getChannelName(contact);
+    return this.chatClient.getUserChannelDescriptors().then((paginator) => {
+      // If this user has channels already, check if there is a channel
+      // between current user and the user being messaged
+      const channel = paginator.items.find(currentChannel => currentChannel.uniqueName === channelName);
+      if (channel) {
+        return channel.getChannel();
+      }
+      return this.createChannelWithUser();
+    });
+  }
+
+  async loadTwilioClient() {
+    let localToken = null; // await AsyncStorage.getItem('twilio_token');
+
+    if (localToken) {
+      const clientReady = await this.initChatClient(localToken);
+      if (clientReady) {
+        return true;
+      }
+    }
+
+    localToken = await this.getTwilioToken();
+    if (localToken) {
+      await AsyncStorage.setItem('twilio_token', localToken);
+      const clientInit = await this.initChatClient(localToken);
+      return clientInit;
+    }
+
+    return false;
+  }
+
+  async getTwilioToken() {
+    const localToken = JSON.parse(this.hmToken);
+    const requestUri = `${API_URL}/chat/token/${localToken._id}?token=${localToken.api_key}`;
+    // TODO: add device ID
+    const bodyData = { device: 'test' };
+    const response = await fetch(requestUri, { method: 'post', body: JSON.stringify(bodyData), headers: { 'Content-Type': 'application/json' } });
+    const responseJson = await response.json();
+    const twilioToken = responseJson.chat_token;
+    return twilioToken;
+  }
+
+  getChannelName(contact) {
+    const localToken = JSON.parse(this.hmToken);
+    // const contact = this.props.navigation.state.params.mentee._id;
+    return localToken._id > contact ? `${localToken._id}.${contact}` : `${contact}.${localToken._id}`;
+  }
+
+  async updateMessages(contact) {
+    const messages = await this.getRawMessages(contact);
+    await this.messageService.updateLocalMessageState(contact, messages);
+  }
+
+  async getRawMessages(contact) {
+    try {
+      const c = await this.channels[contact];
+      return c.getMessages();
+    } catch (e) {
+      // TODO: add sentry logging
+    }
+    return null;
+  }
+
+  async tryInviteContact(contact) {
+    try {
+      // TODO: if the user isn't found, or if the user trying to invite doesn't have permission, we still
+      //  end up showing the error screen, and not catching the failure...
+      const c = await this.channels[contact];
+      c.invite(contact);
+    } catch (e) {
+      // TODO: add sentry logging
+    }
+  }
+
+  async createChannelWithUser(contact) {
+    const channelName = this.getChannelName(contact);
+    return this.chatClient.createChannel({
+      uniqueName: channelName,
+      friendlyName: channelName,
```

Comment:
Duplicate values here, not sure if this was intended

Reply:
Yeah, Twilio requires both a uniqueName and friendlyName for each channel. We only care about the uniqueName for now, so we use the single `channelName` for both.